src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -734,6 +783,144 @@ AddToAttributeEquivalenceClass(AttributeEquivalenceClass **attributeEquivalanceC
 
 
 /*
+ * AddUnionSetOperationsToAttributeEquivalenceClass recursively iterates on all the
+ * setOperations and adds each corresponding target entry to the given equivalence
+ * class.
+ */
+static void
+AddUnionSetOperationsToAttributeEquivalenceClass(Query *query,
+												 SetOperationStmt *setOperation,
+												 Var *varToBeAdded,
+												 AttributeEquivalenceClass **
+												 attributeEquivalenceClass)
+{
+	Node *leftArgument = setOperation->larg;
+	Node *rightArgument = setOperation->rarg;
+
+	/* we're only interested in UNIONs */
+	if (setOperation->op != SETOP_UNION)
+	{
+		return;
+	}
+
+	if (IsA(leftArgument, RangeTblRef))
```

Comment:
We are working on a tree structure and adding qualified items to a list like structure. Separating search from actual processing may simplify the code. Right now we are evaluating 4 different options (left-rangetblref, left-setoperationstmt, right-rangetblref, right-setoperationstmt). What we actually need is to find out rangetblref's and call the processor function.